Add GPU support for POSIX shared memory communication

pavanbalaji opened the pull request:
## Pull Request Description

Add GPU support for POSIX shared memory communication.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

minsii commented on src/mpi/datatype/typerep/src/typerep_yaksa_pack.c:
```diff
@@ -17,6 +17,37 @@ int MPIR_Typerep_pack(const void *inbuf, MPI_Aint incount, MPI_Datatype datatype
+    /* fast-path for contig CPU buffers */
+    bool is_contig;
+    MPIR_Datatype_is_contig(datatype, &is_contig);
+    if (is_contig) {
```
Can we get is_contig and true_lb at the same time? I see two pointer dereferences here if the datatype is not built-in.

pavanbalaji replied:
Hmm?  First of all, there's no dereference for builtin types.  Second, these are macros, so it'd essentially be the same as getting both at the same time.

pavanbalaji replied:
One potential optimization would be to do this only for builtin types, rather than contiguous types.  That way, we can get rid of the pointer dereference for size too.  That'll obviously hurt contiguous derived types, though.

minsii replied:
I was concerning the pointer dereferences for non-built-in continuous datatypes. Why not get is_contig and true_lb at the same time instead of calling each MACRO separately?

minsii replied:
You are right that we can get size both at the same time. For example, using `MPIDI_Datatype_check_contig_size_lb` before the if check.

If it is not contiguous, we don't care the extra store instructions (for size and true_lb) anyway.

minsii replied:
Oh, I just notice that `MPIDI_Datatype_check_contig_size_lb` is ch4 internal. Why do we have two sets of datatype attribute functions, some in MPIR but some others in ch4?

pavanbalaji replied:
@minsii  That's a `ch4` macro.  We cannot use it in `typerep`.

pavanbalaji replied:
Also, my point is that, with macros it'd be equivalent.  The compiler will merge the code segments for us.

minsii replied:
I am not sure if the compiler can detect that the three datatype pointer dereferences are the same. I noticed that the macro is in ch4. But why can't we move it to MPIR if it is useful here?

pavanbalaji replied:
They are not the same.  That's my point.  Which pointer dereferences are the same that you are hoping the compiler can merge if we used a single macro?

minsii replied:
For non-builtin datatypes, I see the is_contig, size, and true_lb attributes are obtained from the following code in your current version:
```
#define MPIR_Datatype_is_contig(dtype_, is_contig_)                            \
        ...
        else {                                                                 \
            MPIR_Datatype *dtp_ = NULL;                                        \
            MPIR_Datatype_get_ptr((dtype_), dtp_);                             \
            MPIR_Assert(dtp_ != NULL);                                         \
            *(is_contig_) = dtp_->is_contig;                                   \


#define MPIR_Datatype_get_size_macro(a,size_) do {                     \
        ...
        case HANDLE_KIND_DIRECT:                                        \
            MPIR_Assert(HANDLE_INDEX(a) < MPIR_DATATYPE_PREALLOC);      \
            ptr = MPIR_Datatype_direct+HANDLE_INDEX(a);                 \
            size_ = ((MPIR_Datatype *) ptr)->size;                      \
            break;                                                      \

#define MPIR_Datatype_get_true_lb(dtype_, true_lb_)                            \
        ...
        else {                                                                 \
            MPIR_Datatype *dtp_ = NULL;                                        \
            MPIR_Datatype_get_ptr((dtype_), dtp_);                             \
            *(true_lb_) = dtp_->true_lb;                                       \
```

Can they be combined similar to `MPIDI_Datatype_check_contig_size_lb `? Please see code capture below. Then the `MPIR_Datatype_get_ptr` will be called only once.
```
            MPIR_Datatype *dt_ptr_;                             \
            MPIR_Datatype_get_ptr((datatype_), (dt_ptr_));      \
            (dt_contig_out_) = (dt_ptr_)->is_contig;        \
            (data_sz_out_)   = (size_t)(count_) *           \
                (dt_ptr_)->size;                            \
            (dt_true_lb_)    = (dt_ptr_)->true_lb;          \
```

pavanbalaji replied:
Your final "optimized" example still has the three pointer deferences:

```
            (dt_contig_out_) = (dt_ptr_)->is_contig;        \
            (data_sz_out_)   = (size_t)(count_) *           \
                (dt_ptr_)->size;                            \
            (dt_true_lb_)    = (dt_ptr_)->true_lb;          \
```

Where are you saving pointer dereferences?

minsii replied:
Oops, sorry. Wrong term here. I meant the computation for getting the pointer of the datatype. I.e.,
```
        case HANDLE_KIND_DIRECT:                                        \
            ptr=MPIR_##kind##_direct+HANDLE_INDEX(a);                   \
```

pavanbalaji replied:
I see.  That's not a pointer dereference, though.  That's just an arithmetic operation.  I guess it is possible to save that too, but that really looks like a stretch at this point.  :-)

I'll workaround it in my code, because I don't want to argue about it anymore.  :-)

pavanbalaji replied:
Is this new version better?

pavanbalaji replied:
Yikes, hold off on the review.  I tried to combine the changes to a single function, but screwed up which buffer the true_lb should apply to.

minsii commented on src/mpi/datatype/typerep/src/typerep_yaksa_pack.c:
```diff
@@ -46,6 +77,37 @@ int MPIR_Typerep_unpack(const void *inbuf, MPI_Aint insize, void *outbuf, MPI_Ai
+            goto fn_exit;
+        }
+    }
+
```
Can we wrap the same contig copy code into a separate routine and reuse in both the pack and unpack functions?

pavanbalaji replied:
We can't.  They are not exactly the same.  For example, the true_lb is applied to different buffers in the two cases.  The offset applies to different buffers.

minsii commented on src/mpi/datatype/typerep/src/typerep_yaksa_pack.c:
```diff
@@ -17,6 +62,21 @@ int MPIR_Typerep_pack(const void *inbuf, MPI_Aint incount, MPI_Datatype datatype
+
+    /* fast-path for contig CPU buffers */
+    if (inoffset == 0) {
+        bool fast = fastpath_memcpy(inbuf, outbuf, datatype, incount, max_pack_bytes,
```
Why do we try fastpath only if `inoffset==0`?

pavanbalaji replied:
That was one of the approximations I had to make to try to put the code in a single function.  The reasoning was that, if in offset is not zero it's a large message anyway, so the optimization shouldn't matter much.

minsii replied:
Can we pass `inbuf + inoffset` to the subroutine instead of `inbuf`? I see `inbuf` is always used with the offset in the original code. Same for outbuf and outoffset

pavanbalaji replied:
No, the code is incorrect -- see my comment above.  Fixing it.  But, in order to fix it, I'll need to pass the copy direction (pack vs. unpack) to the function.  So, I might as well pass the offset in there and decide whether to apply that to the inbuf or outbuf inside the function.

pavanbalaji replied:
Pushed a new version.  I have done no local testing (not even warning testing).  Sorry.  Maybe wait for a few minutes for jenkins to do the warning testing?

minsii replied:
I assume that you are fixing the warning errors. I will wait for you updates.

pavanbalaji replied:
Pushed another version.  Tested locally this time.  :-)

pavanbalaji replied:
@minsii Can you take a look at this again?

minsii replied:
Yes, I was looking at it. It looks good to me.

pavanbalaji commented:
test:mpich/ch4/most

pavanbalaji commented:
Minor bug fix based on the testing.  No need to review again.

test:mpich/ch4/most

pavanbalaji commented:
test:mpich/ch4/most

pavanbalaji commented:
This PR was a mistake.  We already know contiguous types in CH4.  When we converted `MPIR_Memcpy` to `MPIR_Typerep_pack`, we lost the contig/noncontig information and converted a simple byte-based copy operation to a datatype pack operation -- this unnecessarily adds type conversion overhead, when it was not needed at all.  It was a very silly design on my part.

The right approach would have been to create a new function called `MPIR_Typerep_copy`, which basically takes the same parameters as `MPIR_Memcpy`, but either calls `MPIR_Memcpy` or `yaksa` depending on whether the data is on the CPU or the GPU.
